### PR TITLE
[language] Native Functions Refactor

### PR DESCRIPTION
**Pull request by dariorussi**
This is mostly removing a verification step for native functions that
was not adding anything in terms of safety but may have had some value
in making developers think harder about native functions.
It does not feel the effort, code, maintenence is giving us value.
We should implement at some point signature checks in the proper places
so we have better protection than ever, but for now this is not less
secure than what we have.
The simplification also makes it easier to split the code so that
the declaration of what native functions are accepted now lives inside
the MoveVM.
There is more coming and this is about changes in the
way we do verification/resolution/dispatching.
I think it needs its own PR

### CODE REVIEW

File: language/bytecode-verifier/src/verifier.rs [hunk 1]
```diff
@@ -234,105 +231,6 @@ pub fn verify_script_dependency_map(
     verify_dependencies(&fake_module, dependency_map)
 }
 
-fn verify_native_functions(module_view: &ModuleView<VerifiedModule>) -> VMResult<()> {
```
Comment (dariorussi) [inline language/bytecode-verifier/src/verifier.rs:237]: part of this is gone forever and the part that matters and was used at runtime is in the loader

File: language/move-vm/runtime/src/interpreter.rs [hunk 2]
```diff
@@ -235,22 +234,13 @@ impl<'txn> Interpreter<'txn> {
...
-        let function_name = function.name();
-        let native_function =
-            FunctionResolver::resolve(module_id, function_name).ok_or_else(|| {
-                VMStatus::new(StatusCode::LINKER_ERROR)
-                    .with_message(format!("Cannot find native function {}", function_name))
-            })?;
-
         let mut arguments = VecDeque::new();
-        let expected_args = native_function.num_args();
+        let expected_args = function.arg_count();
```
Comment (dariorussi) [inline language/move-vm/runtime/src/interpreter.rs:238]: no reason to ask the native function, in fact it's risky that if there was a mismatch we would not push or pop correct number of values into/from

File: language/move-vm/runtime/src/loader.rs [hunk 3]
```diff
@@ -414,16 +429,40 @@ impl Loader {
...
     }
 
-    fn load_module(
-        &self,
-        id: &ModuleId,
-        context: &dyn InterpreterContext,
-    ) -> VMResult<Arc<Module>> {
-        if let Some(module) = self.module_cache.lock().unwrap().get(id) {
-            return Ok(module);
+    pub(crate) fn check_natives(module: &VerifiedModule) -> VMResult<()> {
```
Comment (dariorussi) [inline language/move-vm/runtime/src/loader.rs:432]: this is the "linker" check. 
It checks this VM is exposing the given function as native.
In time we can use the signature in `Function` to verify arguments

File: language/move-vm/runtime/src/native_functions.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Libra Core Contributors
```
Comment (dariorussi) [inline language/move-vm/runtime/src/native_functions.rs:1]: native functions exposed live inside the Move VM and the are private to the VM.
The function themselves can come from different crates, but the MoveVM controls what it wants to expose

File: language/move-vm/types/src/native_functions/dispatch.rs [hunk 5]
```diff
@@ -1,32 +1,12 @@
 // Copyright (c) The Libra Core Contributors
```
Comment (dariorussi) [inline language/move-vm/types/src/native_functions/dispatch.rs:1]: dispatch is left with utilities to implement native functions

**Comment by dariorussi** [comment]
for what is worth and for a good laugh removing the resolution at runtime improves performance of a simple test calling natives of ~5%.
Given the number of native functions and the optimized hash map implemented in resolution (a complex switch) that seems a good win :) 
Anyway just mentioning that things did not get worse

File: language/bytecode-verifier/src/verifier.rs (hunk 1 above)
Comment (tnowacki) [reply language/bytecode-verifier/src/verifier.rs:237]: We should still verify that the native function exists, no?

File: language/bytecode-verifier/src/verifier.rs [hunk 6]
```diff
@@ -234,105 +231,6 @@ pub fn verify_script_dependency_map(
...
-                    ));
-                }
-            }
-        }
-    }
-    Ok(())
-}
-
-// TODO: native structs have been partially removed. Revisit.
-fn verify_native_structs(module_view: &ModuleView<VerifiedModule>) -> VMResult<()> {
```
Comment (tnowacki) [inline language/bytecode-verifier/src/verifier.rs:321]: same here, I don't feel like this should go away

File: language/move-vm/runtime/src/loader.rs (hunk 3 above)
Comment (tnowacki) [reply language/move-vm/runtime/src/loader.rs:432]: Maybe my confusion comes from when this is called. When is this invoked? Every time the module is loaded?

File: language/bytecode-verifier/src/verifier.rs (hunk 1 above)
Comment (dariorussi) [reply language/bytecode-verifier/src/verifier.rs:237]: not in the verifier, how would the verifier know? the VM is the only one that knows, no?

File: language/move-vm/runtime/src/loader.rs (hunk 3 above)
Comment (dariorussi) [reply language/move-vm/runtime/src/loader.rs:432]: every time the module is loaded and when a module is published.
I though I explained it but I see now I only did to myself ;)
- On publish, if a module has a native the VM/Runtime is asked if that native is expected. If not the module does not verify and does not publish. Implication is also that to add a native you have to version the VM before you can push the native
- On loading, a native is linked, that is, is resolved and the "function" cached away, so linked at load time as it should be.
The hope is that the combination of both gives us enough guarantees to feel good.

Linkage can and will be extended to some signature check. We can talk about ways to accomplish that, I have some idea but don't know if good or bad

File: language/bytecode-verifier/src/verifier.rs (hunk 6 above)
Comment (dariorussi) [reply language/bytecode-verifier/src/verifier.rs:321]: this follows the same logic as the functions. It's a publish and load time.
However structs in this PR are shortcircuit'ed in a horrible way and we should talk about where to fail them.
In other words there is nothing a verifier - or anything that is not the runtime that owns the natives - can do or decide upon. 
To put it another way, the verifier should not "version" if natives change.

Thoughts?

File: language/move-vm/runtime/src/loader.rs (hunk 3 above)
Comment (dariorussi) [reply language/move-vm/runtime/src/loader.rs:432]: oh also, that is basically what happens now with the check in the verifier.
We are just making all the decision and semantic of natives part of the VM and not the verifier

File: language/bytecode-verifier/src/verifier.rs (hunk 1 above)
Comment (tnowacki) [reply language/bytecode-verifier/src/verifier.rs:237]: I understand how things are laid out now

File: language/bytecode-verifier/src/verifier.rs (hunk 6 above)
Comment (tnowacki) [reply language/bytecode-verifier/src/verifier.rs:321]: This makes sense. I think what is there in the new `check_natives` is good for now

**Comment by dariorussi** [comment]
@tnowacki, @vgao1996 let me know if you have changes or things you'd like to see happening here. I got more PRs coming in this space...

**Comment by vgao1996** [comment]
@dariorussi just FYI it looks good to me! I had some concerns but those were already addressed in our meeting.

**Pull request closed without merging**